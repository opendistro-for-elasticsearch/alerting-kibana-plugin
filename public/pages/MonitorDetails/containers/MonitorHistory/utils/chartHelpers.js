/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { ALERT_STATE } from '../../../../../utils/constants';
import { TIME_SERIES_ALERT_STATE } from './constants';

export const dataPointsGenerator = ({
  startTime,
  acknowledgedTime,
  endTime,
  lastEndTime,
  windowEndTime,
  meta,
}) => {
  const alertDataPoints = [];

  // Adds not alerting state
  if (lastEndTime) {
    alertDataPoints.push({
      x0: lastEndTime,
      x: startTime,
      state: TIME_SERIES_ALERT_STATE.NO_ALERTS,
      meta: {
        startTime: lastEndTime,
        endTime: startTime,
      },
    });
  }

  if (acknowledgedTime) {
    alertDataPoints.push({
      x0: startTime,
      x: acknowledgedTime < windowEndTime ? acknowledgedTime : windowEndTime,
      state: TIME_SERIES_ALERT_STATE.TRIGGERED,
      meta,
    });
    // If acknowledge time is outside window, ignore the end plotting
    if (acknowledgedTime < windowEndTime) {
      alertDataPoints.push({
        x0: acknowledgedTime,
        x: endTime && endTime < windowEndTime ? endTime : windowEndTime,
        state: TIME_SERIES_ALERT_STATE.ACKNOWLEDGED,
        meta,
      });
    }
  } else {
    alertDataPoints.push({
      x0: startTime,
      x: endTime && endTime < windowEndTime ? endTime : windowEndTime,
      state: TIME_SERIES_ALERT_STATE.TRIGGERED,
      meta,
    });
  }
  return alertDataPoints;
};

export const generateFirstDataPoints = ({
  startTime,
  acknowledgedTime,
  endTime,
  state,
  windowStartTime,
  windowEndTime,
  errorsCount,
}) => {
  const firstAlertData = {
    startTime,
    acknowledgedTime,
    endTime,
    windowStartTime,
    windowEndTime,
    meta: {
      startTime,
      acknowledgedTime,
      endTime,
      state,
      errorsCount,
    },
  };

  /*
    Alert could have started within timeSeriesWindowRange,
    in this case , it should create NOT_ALERTING data point
    for startWindow -> firstAlertTime and then continue creating
    alerts
  */
  if (windowStartTime < firstAlertData.startTime) {
    return dataPointsGenerator({
      ...firstAlertData,
      lastEndTime: windowStartTime,
    });
  }
  // Alert is being continued from previous window, highlight from
  // startOfWindow to end_time / Acknowledge Time
  return dataPointsGenerator({
    ...firstAlertData,
    startTime: windowStartTime,
    acknowledgedTime:
      acknowledgedTime && acknowledgedTime < windowStartTime ? windowStartTime : acknowledgedTime,
    lastEndTime: null,
  });
};

// This generates query for the displaying POI Data from the window with provided duration
export const getPOISearchQuery = (monitorId, startTime, endTime, intervalDuration) => ({
  query: {
    bool: {
      filter: [
        { term: { monitor_id: monitorId } },
        {
          bool: {
            should: [
              { range: { start_time: { gte: startTime, lte: endTime } } },
              { range: { end_time: { gte: startTime, lte: endTime } } },
              {
                bool: {
                  should: [
                    { range: { start_time: { lte: startTime } } },
                    { range: { end_time: { gte: endTime } } },
                    {
                      bool: {
                        must_not: {
                          exists: { field: 'monitor.schedule.period' },
                        },
                      },
                    },
                  ],
                  minimum_should_match: 2,
                },
              },
            ],
            minimum_should_match: 1,
          },
        },
      ],
    },
  },
  aggs: {
    alerts_over_time: {
      date_histogram: {
        interval: `${intervalDuration.asMinutes()}m`,
        extended_bounds: {
          min: startTime,
          max: endTime,
        },
        script: {
          lang: 'painless',
          // Alert start / end time can fall in only in one bucket, but our POI shows the alerts over the time of period.
          // To achieve this below script will return events over the period so it will fall into all buckets from starting to end time for an alert
          source: `def alertsWindow = new ArrayList();
                      def startTime = doc.start_time.value.millis < params.startTime ? params.startTime : doc.start_time.value.millis;
                      def endTime;
                      /*
                        Painless script returns default dateTime epoch for null date fields instead of returning null.
                        In case of alert is in one of the state (ACTIVE / ERROR / ACKNOWLEDGE) end_time can be null.
                        We are setting default end time to window end time.
                        More details can be found here https://github.com/elastic/elasticsearch/pull/30975
                      */

                      if(doc.end_time.size() == 0) {
                        endTime = params.endTime
                      } else {
                        endTime = doc.end_time.value.millis;
                      }
                      while(startTime <= endTime) {
                           alertsWindow.add(startTime);
                           startTime = startTime + params.interval;
                       }
                       return alertsWindow;`,
          params: {
            startTime,
            endTime,
            interval: intervalDuration.asMilliseconds(),
          },
        },
      },
    },
    max_alerts: {
      max_bucket: {
        buckets_path: 'alerts_over_time._count',
      },
    },
  },
});
