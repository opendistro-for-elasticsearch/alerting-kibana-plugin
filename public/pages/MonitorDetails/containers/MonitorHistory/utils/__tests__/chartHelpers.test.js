/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import moment from 'moment-timezone';
import { generateFirstDataPoints, dataPointsGenerator, getPOISearchQuery } from '../chartHelpers';

moment.tz.setDefault('America/Los_Angeles');

describe('Chart Helpers', () => {
  describe('getPOISearchQuery', () => {
    it('Generates SearchQuery for the POI Graph', () => {
      const startTime = moment('2018-10-29T09:00:00').valueOf();
      const endTime = moment('2018-10-29T10:00:00').valueOf();
      const intervalDuration = moment.duration(10, 'm');
      expect(
        getPOISearchQuery('monitorId', startTime, endTime, intervalDuration)
      ).toMatchSnapshot();
    });
  });
  describe('Time Series', () => {
    const windowTime = {
      startTime: moment('2018-10-29T09:00:00').valueOf(),
      endTime: moment('2018-10-29T10:00:00').valueOf(),
    };
    const alertInputData = {
      startTime: moment('2018-10-29T09:15:00').valueOf(),
      acknowledgedTime: moment('2018-10-29T09:17:00').valueOf(),
      endTime: moment('2018-10-29T09:18:00').valueOf(),
      windowEndTime: windowTime.endTime,
    };
    describe('generateFirstDataPoints should generate correct data point for an alert', () => {
      test('started inside window range', () => {
        expect(
          generateFirstDataPoints({
            ...alertInputData,
            state: 'COMPLETED',
            windowStartTime: windowTime.startTime,
            windowEndTime: windowTime.endTime,
            errorsCount: 0,
          })
        ).toMatchSnapshot();
      });
      test('started outside window range', () => {
        expect(
          generateFirstDataPoints({
            ...alertInputData,
            startTime: moment('2018-10-29T08:55:00').valueOf(),
            state: 'COMPLETED',
            windowStartTime: windowTime.startTime,
            windowEndTime: windowTime.endTime,
            errorsCount: 5,
          })
        ).toMatchSnapshot();
      });
    });
    describe('dataPointsGenerator should generate correct data point for an alert', () => {
      it('inside window range', () => {
        const dataPoints = dataPointsGenerator({
          ...alertInputData,
          acknowledgedTime: null,
          meta: {
            metaKey: 'metaValue',
          },
        });
        expect(dataPoints).toMatchSnapshot();
      });

      it('has acknowledged_time', () => {
        const dataPoints = dataPointsGenerator({
          ...alertInputData,
          meta: {
            metaKey: 'metaValue',
          },
        });
        expect(dataPoints).toMatchSnapshot();
      });

      it('has an acknowledged_time outside window range', () => {
        const dataPoints = dataPointsGenerator({
          ...alertInputData,
          acknowledgedTime: moment('2018-10-29T10:01:00').valueOf(),
          meta: {
            metaKey: 'metaValue',
          },
        });
        expect(dataPoints).toMatchSnapshot();
      });

      it('has an end_time outside window range', () => {
        const dataPoints = dataPointsGenerator({
          ...alertInputData,
          acknowledgedTime: null,
          endTime: moment('2018-10-29T10:01:00').valueOf(),
          meta: {
            metaKey: 'metaValue',
          },
        });
        expect(dataPoints).toMatchSnapshot();
      });

      it('if previous end time has been provided generate no_alert state', () => {
        const dataPoints = dataPointsGenerator({
          ...alertInputData,
          lastEndTime: windowTime.startTime,
          meta: {
            metaKey: 'metaValue',
          },
        });
        expect(dataPoints).toMatchSnapshot();
      });
    });
  });
});
