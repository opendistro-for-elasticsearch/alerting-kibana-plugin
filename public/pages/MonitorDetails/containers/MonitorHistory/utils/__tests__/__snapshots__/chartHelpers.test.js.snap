// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart Helpers Time Series dataPointsGenerator should generate correct data point for an alert has acknowledged_time 1`] = `
Array [
  Object {
    "meta": Object {
      "metaKey": "metaValue",
    },
    "state": "TRIGGERED",
    "x": 1540829820000,
    "x0": 1540829700000,
  },
  Object {
    "meta": Object {
      "metaKey": "metaValue",
    },
    "state": "ACKNOWLEDGED",
    "x": 1540829880000,
    "x0": 1540829820000,
  },
]
`;

exports[`Chart Helpers Time Series dataPointsGenerator should generate correct data point for an alert has an acknowledged_time outside window range 1`] = `
Array [
  Object {
    "meta": Object {
      "metaKey": "metaValue",
    },
    "state": "TRIGGERED",
    "x": 1540832400000,
    "x0": 1540829700000,
  },
]
`;

exports[`Chart Helpers Time Series dataPointsGenerator should generate correct data point for an alert has an end_time outside window range 1`] = `
Array [
  Object {
    "meta": Object {
      "metaKey": "metaValue",
    },
    "state": "TRIGGERED",
    "x": 1540832400000,
    "x0": 1540829700000,
  },
]
`;

exports[`Chart Helpers Time Series dataPointsGenerator should generate correct data point for an alert if previous end time has been provided generate no_alert state 1`] = `
Array [
  Object {
    "meta": Object {
      "endTime": 1540829700000,
      "startTime": 1540828800000,
    },
    "state": "NO_ALERTS",
    "x": 1540829700000,
    "x0": 1540828800000,
  },
  Object {
    "meta": Object {
      "metaKey": "metaValue",
    },
    "state": "TRIGGERED",
    "x": 1540829820000,
    "x0": 1540829700000,
  },
  Object {
    "meta": Object {
      "metaKey": "metaValue",
    },
    "state": "ACKNOWLEDGED",
    "x": 1540829880000,
    "x0": 1540829820000,
  },
]
`;

exports[`Chart Helpers Time Series dataPointsGenerator should generate correct data point for an alert inside window range 1`] = `
Array [
  Object {
    "meta": Object {
      "metaKey": "metaValue",
    },
    "state": "TRIGGERED",
    "x": 1540829880000,
    "x0": 1540829700000,
  },
]
`;

exports[`Chart Helpers Time Series generateFirstDataPoints should generate correct data point for an alert started inside window range 1`] = `
Array [
  Object {
    "meta": Object {
      "endTime": 1540829700000,
      "startTime": 1540828800000,
    },
    "state": "NO_ALERTS",
    "x": 1540829700000,
    "x0": 1540828800000,
  },
  Object {
    "meta": Object {
      "acknowledgedTime": 1540829820000,
      "endTime": 1540829880000,
      "errorsCount": 0,
      "startTime": 1540829700000,
      "state": "COMPLETED",
    },
    "state": "TRIGGERED",
    "x": 1540829820000,
    "x0": 1540829700000,
  },
  Object {
    "meta": Object {
      "acknowledgedTime": 1540829820000,
      "endTime": 1540829880000,
      "errorsCount": 0,
      "startTime": 1540829700000,
      "state": "COMPLETED",
    },
    "state": "ACKNOWLEDGED",
    "x": 1540829880000,
    "x0": 1540829820000,
  },
]
`;

exports[`Chart Helpers Time Series generateFirstDataPoints should generate correct data point for an alert started outside window range 1`] = `
Array [
  Object {
    "meta": Object {
      "acknowledgedTime": 1540829820000,
      "endTime": 1540829880000,
      "errorsCount": 5,
      "startTime": 1540828500000,
      "state": "COMPLETED",
    },
    "state": "TRIGGERED",
    "x": 1540829820000,
    "x0": 1540828800000,
  },
  Object {
    "meta": Object {
      "acknowledgedTime": 1540829820000,
      "endTime": 1540829880000,
      "errorsCount": 5,
      "startTime": 1540828500000,
      "state": "COMPLETED",
    },
    "state": "ACKNOWLEDGED",
    "x": 1540829880000,
    "x0": 1540829820000,
  },
]
`;

exports[`Chart Helpers getPOISearchQuery Generates SearchQuery for the POI Graph 1`] = `
Object {
  "aggs": Object {
    "alerts_over_time": Object {
      "date_histogram": Object {
        "extended_bounds": Object {
          "max": 1540832400000,
          "min": 1540828800000,
        },
        "interval": "10m",
        "script": Object {
          "lang": "painless",
          "params": Object {
            "endTime": 1540832400000,
            "interval": 600000,
            "startTime": 1540828800000,
          },
          "source": "def alertsWindow = new ArrayList();
                      def startTime = doc.start_time.value.millis < params.startTime ? params.startTime : doc.start_time.value.millis;
                      def endTime;
                      /*
                        Painless script returns default dateTime epoch for null date fields instead of returning null.
                        In case of alert is in one of the state (ACTIVE / ERROR / ACKNOWLEDGE) end_time can be null.
                        We are setting default end time to window end time.
                        More details can be found here https://github.com/elastic/elasticsearch/pull/30975
                      */

                      if(doc.end_time.size() == 0) {
                        endTime = params.endTime
                      } else {
                        endTime = doc.end_time.value.millis;
                      }
                      while(startTime <= endTime) {
                           alertsWindow.add(startTime);
                           startTime = startTime + params.interval;
                       }
                       return alertsWindow;",
        },
      },
    },
    "max_alerts": Object {
      "max_bucket": Object {
        "buckets_path": "alerts_over_time._count",
      },
    },
  },
  "query": Object {
    "bool": Object {
      "filter": Array [
        Object {
          "term": Object {
            "monitor_id": "monitorId",
          },
        },
        Object {
          "bool": Object {
            "minimum_should_match": 1,
            "should": Array [
              Object {
                "range": Object {
                  "start_time": Object {
                    "gte": 1540828800000,
                    "lte": 1540832400000,
                  },
                },
              },
              Object {
                "range": Object {
                  "end_time": Object {
                    "gte": 1540828800000,
                    "lte": 1540832400000,
                  },
                },
              },
              Object {
                "bool": Object {
                  "minimum_should_match": 2,
                  "should": Array [
                    Object {
                      "range": Object {
                        "start_time": Object {
                          "lte": 1540828800000,
                        },
                      },
                    },
                    Object {
                      "range": Object {
                        "end_time": Object {
                          "gte": 1540832400000,
                        },
                      },
                    },
                    Object {
                      "bool": Object {
                        "must_not": Object {
                          "exists": Object {
                            "field": "monitor.schedule.period",
                          },
                        },
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
      ],
    },
  },
}
`;
